pkg/env/env.go

```diff
@@ -387,38 +433,116 @@ func (e *testEnv) getFinishActions() []action {
 	return e.getActionsByRole(roleFinish)
 }
 
-func (e *testEnv) execFeature(ctx context.Context, t *testing.T, featName string, f types.Feature) context.Context {
-	// feature-level subtest
-	t.Run(featName, func(t *testing.T) {
-		// skip feature which matches with --skip-feature
-		if e.cfg.SkipFeatureRegex() != nil && e.cfg.SkipFeatureRegex().MatchString(featName) {
-			t.Skipf(`Skipping feature "%s": name matched`, featName)
-		}
-
-		// skip feature which does not match with --feature
-		if e.cfg.FeatureRegex() != nil && !e.cfg.FeatureRegex().MatchString(featName) {
-			t.Skipf(`Skipping feature "%s": name not matched`, featName)
+func (e *testEnv) runTests() bool {
+	return !e.cfg.DryRunMode() && !e.gatherMode
+}
+
+func (e *testEnv) shoulFeatureOrAssessmentBeRun(f types.Feature, a types.Step, assessmentIndex int) (skipped bool, message string) {
```

Comment:
This method looks a little convoluted. I think part of the issue is (maybe) because it is also collecting stats for summary (see my suggestion for removing stats for now). Also, I would suggest delaying making the decision, of figuring out whether an action/feature should be executed, at this point.  Instead, consider pushing that decision futrther down the execution stack [here](https://github.com/kubernetes-sigs/e2e-framework/blob/319e09e8b645bfbae7fc332a32588882e693c813/pkg/env/action.go#L94)

Reply:
Let me move the stats part of this out to a different PR. 

Let me refactor the other change like you suggested into the execution part. I will push an update for this in a day or so.